jparse/man/man1/jfmt.1
.\" section 1 man page for jfmt
.\"
.\" This man page was first written by Cody Boone Ferguson for the IOCCC
.\" in 2023.
.\"
.\" Humour impairment is not virtue nor is it a vice, it's just plain
.\" wrong: almost as wrong as JSON spec mis-features and C++ obfuscation! :-)
.\"
.\" "Share and Enjoy!"
.\"     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
.\"
.TH jfmt 1 "06 August 2023" "jfmt" "IOCCC tools"
.SH NAME
.B jfmt
\- IOCCC JSON printer
.SH SYNOPSIS
.B jfmt
.RB [\| \-h \|]
.RB [\| \-V \|]
.RB [\| \-v
.IR level \|]
.RB [\| \-J
.IR level \|]
.RB [\| \-q \|]
.RB [\| \-L
.IR <num>{[t|s]} \|]
.RB [\| \-I
.IR <num>{[t|s]} \|]
.RB [\| \-l
.IR lvl \|]
.RB [\| \-m
.IR depth \|]
.RB [\| \-K \|]
.RB [\| \-o \|]
.RB [\| \-F
.IR fmt \|]
.IR file.json
.SH DESCRIPTION
.B jfmt
parses a JSON file,
.IR file.json ,
and if it is valid it will format the JSON file in a way that is easier to see the JSON structure.
One might call this pretty-printing or beautifying the JSON file (for various definitions of beautifying JSON :-) ).
By default the tool writes the output to stdout but if you use the
.B \-o
option you can change this behaviour (\- means stdout).
The
.B \-F
option allows you to change the JSON output format.
The
.B \-m
option allows you to change the default depth to traverse when parsing the JSON where the default is set to a reasonable value for most cases (256 or the macro
.BR JSON_DEFAULT_MAX_DEPTH ).
.PP
The option
.B \-l
allows you to restrict the output based on the JSON level.
The option
.B \-L
tells the program to print the JSON level followed by a given number of spaces or tabs where the default is 4 spaces.
.B \-I
allow for indenting the levels with a default value of 4 spaces.
As these options are more complicated in form we describe them more in the
.B OPTIONS
section below.
.SH OPTIONS
.TP
.B \-h
Show help and exit.
.TP
.B \-V
Show version and exit.
.TP
.BI \-v\  level
Set verbosity level to
.IR level
(def: 0).
.TP
.BI \-J\  level
Set JSON verbosity level to
.IR level
(def: 0).
.TP
.B \-q
Suppress some of the output (def: show more information).
.TP
.BI \-L\   <num>[{t|s}
Print the JSON levels followed by the number of spaces or tabs, depending on the letter specified,
.BR t
for tabs and
.BR s
for spaces.
Not specifying either defaults to spaces.
As a convenience, specifying
.BI \-L\  tab
is an alias for
.BI \-L\  1t\c
\&.
.TP
.BI \-I\   <num>[{t|s}
Indent the JSON levels by the number of spaces or tabs, depending on the letter specified,
.BR t
for tabs and
.BR s
for spaces.
Not specifying either defaults to spaces.
As a convenience, specifying
.B \-I\  tab
is an alias for
.BI \-I\  1t\c
\&.
.TP
.BI \-l\  lvl
Only show JSON members in the given level range (def: show all levels).
.sp
.RS
If
.I lvl
is a number (e.g.: \-l 3), level must == number.
.br
If
.I lvl
is a number followed by : (e.g. '\-l 3:'), level must be >= number.
.br
If
.I lvl
is a : followed by a number (e.g. '\-l :3'), level must be <= number.
.br
If
.I lvl
is num:num (e.g. '\-l 3:5'), level must be inclusively in the range.
.RE
.TP
.BI \-m\  max_depth
Set maximum traversal depth when parsing the JSON document.
The default is defined as the macro
.BI JSON_DEFAULT_MAX_DEPTH ,
256.
Specifying 0 means infinite depth: only safe with infinite time, variable size and RAM! :-)
.TP
.B \-K
Run built\-in test suite of some features.
.TP
.BI \-o\  ofile
Write output to specified file.
.RS
If the file already exists (case-insensitive) it is an error.
.br
If the file cannot be opened for writing it is also an error.
.br
Default is stdout which is the same as if you specified
.BR \- .
.RE
.TP
.BI \-F\  fmt
Change the JSON output format of the tool.
.RS
.TQ
.B tty
When output is to a TTY, use
.BR colour ,
otherwise use
.BR simple .
.RE
.RS
.TQ
.B simple
Each line has one JSON level determined by '[]'s and '{}'s
.RE
.RS
.TQ
.B colour
.B simple
plus ANSI colour syntax highlighting
.RE
.RS
.TQ
.B color
Alias for
.B colour
.RB (  colour
excluding yo(\c
.BR u )
:\-) ), see below subsection
.RS
.RE
.RE
.RS
.TQ
.B 1line
One line output
.RE
.RS
.TQ
.B nows
One line output, no extra whitespace
.RE
.SS Coloured output
Although we do allow for
.B color
instead of
.B colour
it should be noted that this is not only strictly incorrect but it also prevents it from being used by both you and me.
In particular it excludes yo(\c
.BR u )
so we recommend that you only use
.B color
if you want to exclude yourself. :\-)
.SH EXIT STATUS
.TP
0
all is OK and file is valid JSON
.TQ
1
error writing to ofile
.TQ
2
\-h and help string printed or \-V and version string printed
.TQ
3
invalid command line, invalid option or option missing an argument
.TQ
4
file does not exist, not a file, or unable to read the file
.TQ
5
file contents is not valid JSON
.TQ
6
test mode failed
.TQ
>=10
internal error
.SH NOTES
.PP
The JSON parser was written as a collaboration between Cody Boone Ferguson and Landon Curt Noll, one of the IOCCC Judges, to support
IOCCCMOCK, IOCCC28 and beyond.
.PP
.BR jfmt (1)
is being developed by Cody Boone Ferguson.
.PP
For more detailed history that goes beyond this humble man page we recommend you check
.BR jfmt (1),
.IR CHANGES.md ,
.IR README.md ,
the GitHub git log as well as reading the source code (or not :\-) ).
.PP
We don't recommend you check the GitHub issue page! :\-)
This is because it's incredibly long with a lot of OT things and would take even the fastest readers a very long time to read. :\-(
.PP
.SH BUGS
It is currently incomplete and listing the missing features and things that are not correct is not worth the time or effort.
.PP
It's important to note that one should not rely on the exact output of this tool and that future versions might change the way JSON is formatted.
.SH EXAMPLES
.PP
Pretty print a JSON file
.I h2g2.json
if it is valid JSON:
.sp
.RS
.ft B
jfmt h2g2.json
.ft R
.RE
.PP
Print the levels of each part of the JSON document
.BR jparse/test_jparse/test_JSON/good/foo.json
followed by 4 spaces:
.sp
.RS
.ft B
jfmt -L 4s jparse/test_jparse/test_JSON/good/foo.json
.ft R
.RE
.sp
This is a useful way to better understand JSON levels.
.SH SEE ALSO
.PP
.BR jfmt (1),
.BR jparse (1),
.BR jval (1),
.BR jnamval (1)


CHANGES.md
# Major changes to the IOCCC entry toolkit

## Release 1.0.52 2023-08-09

New version of `jval` and `jnamval`, `"0.0.14 2023-08-09"`.

Fixed the lists of `-S` and `-n` option parsing. For now it should just be a
string that's strdup()d. The operators are in the list in the order specified,
one list for strings and one list for numbers. This can be changed to a single
list later on if necessary (as I suspect it might need to be but the way it is
now is set up as two which is what I'm operating under).

The function that frees the `-S` and `-n` lists in `jval` and `jnamval` is now
in `json_util.c` as they are actually in a struct common to both `jval` and
`jnamval`. The functions that free them in `jval_util.c` and `jnamval_util.c`
simply check that `jval` or `jnamval` is not NULL and then calls the new
`free_json_util_cmp_list()` function.

Made the json util operator macros an enum.

Rename the enum `output_format` to `JSON_UTIL_OUTPUT_FMT`.


## Release 1.0.51 2023-08-07

Fix link in make rule `jparse.clone` to use
`https://github.com/xexyl/jparse.git` as the link used,
`https://github.com/lcn2/jparse.git`, is actually a clone of the repo I set up
(empty for now) which will be populated with `jparse` later on when the parser
is tested more fully.

Add `-F` to common options of `json_util_README.md`.

Change `jnamval -P` to `jnamval -r` (restrict printing types). There might be a
reference to something non-programming in this change. New version `"0.0.13
2023-08-07"`.

Slightly improve `dbg_example.c` by adding a call to `not_reached()` after the
`errp()` call.

## Release 1.0.50 2023-08-06

New version of `jfmt`, `jval` and `jnamval`: `"0.0.12 2023-08-06"`. `jfmt` and
`jval` jumped from 0.0.10 to make them match. Later on after version 1.0.0 this
will no longer happen.

Updated the `-F` option list. It now is:

```
-f fmt		change the json format style (def: use default)

		tty		when output is to a tty, use colour, otherwise use simple
		simple		each line has one json level determined by '[]'s and '{}'s
		colour		simple plus ansi colour syntax highlighting
		color		alias for colour (colour excluding you :-) )
		1line		one line output
		nows		One line output, no extra whitespace
```

Updated man pages for the three tools for this change as well as a typo fix in
`jnamval.1` for the `-p` option (wrongly listed as `-P` which is something
else).

Updated `tmp/TODO.md`: the options for the three tools should be in the same
order, particularly the common options, both in usage string and man pages. But
since the specs are not finalised this needn't be done yet.


## Release 1.0.49 2023-08-05

New version of `jnamval`: `"0.0.11 2023-08-05"`. Add `-I` option with same
semantics as `jfmt`. Updated `jnamval.1` to add this option and fix typos.

Change available options for `jnamval -t` to be that of `jval -t`. Updated man
page.

Typo fixes in `jfmt.1` and `jval.1`. Removed `-I` from `jval.1` as it's not a
valid option.

Fix typo in usage string in `jnamval` which referred to non-existent option (it
is actually an option arg, not an option, that is correct and that is now what
it says).

Add option `-P` to `jnamval`. This is just like what `-t` for `jnamval` used to
be but it will indicate what to print rather than what to match. Man page
updated for this option. As the code is common to both `jval` and `jnamval` the
`jval` utility could be updated easily to support this option is desired but
this is TBD later.

Add some bad test JSON files in the `jparse/test_jparse/test_JSON/bad_loc`
directory, updated Makefile rule Added appropriate `err` files.
Updated Makefile rule `rebuild_jparse_err_files` to not show the command being
run for each file to make the output clean (and to show what it looks like in
the README.md file - which has been updated due to slight changes).

Fix warning message of `mkiocccentry -y` and move it to happen before the option
to ignore all warnings as that one says no more warnings will be displayed so it
should come last.

Removed the "program" arg from `parse_verbosity()`.  The `parse_verbosity()`
no longer exits on errors or bad arguments.  Instead `parse_verbosity()`
returns `DBG_INVALID`, a value < 0.

The return from `parse_verbosity()` is now checked for a < 0 return
value and an appropriate error is raised when that happens.


## Release 1.0.48 2023-08-04

Bug fixes in `location` tool wrt the `-a` and `-s` options. New version `"1.0.2
2023-08-04"`.

Change `-N` to `-n`.

Always use re-entrant version of the lookup functions as this allows for proper
showing of both name and code whether or not one is using `-s` or `-n`. If `-a`
we still show all (as before) but without `-n` we previously showed both name
and code like `code ==> name` when it should only show both if verbosity level
is > 0. Note that the chkentry tool _SHOULD NOT_ and _DOES NOT_ use the
re-entrant versions.

Use `parse_verbosity()` for parsing the `-v` option.

Updated man page with an example added and updating `-N` to `-n`.


## Release 1.0.47 2023-08-03

New version of `location` tool: `"1.0.1 2023-08-03"`.

Improved tool to allow for `-a` (show all) and `-s` (substring search). Uses new
re-entrant versions of the functions `lookup_location_name()` and
`lookup_location_code()`. These new functions take a `size_t *idx` which is set
to the _next_ element in the table (if != NULL) so that the next call can skip
continue _after_ the previous element. They also take a `struct location **`
which stores the _found_ location. This is useful for `-v 1` as one can then see
something like:


```sh
./location -a -s -N -v 1 'united'
United Arab Emirates ==> AE
United Kingdom of Great Britain and Northern Ireland (the) ==> GB
United States Miscellaneous Pacific Islands ==> PU
Tanzania, United Republic of ==> TZ
United Kingdom ==> UK
United States Minor Outlying Islands ==> UM
United Nations ==> UN
United States of America ==> US
```

Without the `-v 1` it will only show:

```sh
$ ./location -a -s -N 'united'
AE

jparse/jfmt.c
/*
 * jfmt - JSON printer
 *
 * "Because specs w/o version numbers are forced to commit to their original design flaws." :-)
 *
 * This tool is being developed by:
 *
 *	@xexyl
 *	https://xexyl.net		Cody Boone Ferguson
 *	https://ioccc.xexyl.net
 *
 * The JSON parser was co-developed in 2022 by Cody and Landon.
 *
 * "Because sometimes even the IOCCC Judges need some help." :-)
 *
 * "Share and Enjoy!"
 *     --  Sirius Cybernetics Corporation Complaints Division, JSON spec department. :-)
 */

/* special comments for the seqcexit tool */
/* exit code out of numerical order - ignore in sequencing - ooo */
/* exit code change of order - use new value in sequencing - coo */

#include "jfmt.h"

/*
 * definitions
 */
#define REQUIRED_ARGS (1)	/* number of required arguments on the command line */


/*
 * static globals
 */
static bool quiet = false;				/* true ==> quiet mode */

/*
 * usage message
 *
 * Use the usage() function to print the usage_msg([0-9]?)+ strings.
 */
static const char * const usage_msg0 =
    "usage:\t%s [-h] [-V] [-v level] [-J level] [-q] [-L <num>{[t|s]}] [-I <num>{[t|s]}]\n"
    "\t[-l lvl] [-m depth] [-K] [-o ofile] [-F fmt] file.json\n"
    "\n"
    "\t-h\t\tPrint help and exit\n"
    "\t-V\t\tPrint version and exit\n"
    "\t-v level\tVerbosity level (def: %d)\n"
    "\t-J level\tJSON verbosity level (def: %d)\n"
    "\t-q\t\tSuppress some of the output (def: show more info)\n"
    "\n"
    "\t-L <num>[{t|s}]\tPrint JSON level followed by a number of tabs or spaces (def: don't print levels)\n"
    "\t-L tab\t\tAlias for: '-L 1t'\n"
    "\n"
    "\t\t\tTrailing 't' implies <num> tabs whereas trailing 's' implies <num> spaces.\n"
    "\t\t\tNot specifying 's' or 't' implies spaces.\n"
    "\n"
    "\t-I <num>{[t|s]}\tWhen printing JSON syntax, indent levels (def: indent with 4 spaces)\n"
    "\t-I tab\t\tAlias for '-I 1t'\n"
    "\n"
    "\t\t\tTrailing 't' implies indent with number of tabs whereas trailing 's' implies spaces.\n"
    "\t\t\tNot specifying 's' or 't' implies spaces.\n"
    "\n"
    "\t-l lvl\t\tPrint specific JSON levels (def: print at any level)\n"
    "\n"
    "\t\t\tIf lvl is a number (e.g.: -l 3), level must == number.\n"
    "\t\t\tIf lvl is a number followed by : (e.g. '-l 3:'), level must be >= number.\n"
    "\t\t\tIf lvl is a : followed by a number (e.g. '-l :3'), level must be <= number.\n"
    "\t\t\tIf lvl is num:num (e.g. '-l 3:5'), level must be inclusively in the range.\n"
    "\n"
    "\t-m max_depth\tSet the maximum JSON level depth to max_depth (0 == infinite depth, def: %d)\n"
    "\n"
    "\t\t\tA 0 max_depth implies JSON_INFINITE_DEPTH: only safe with infinite variable size and RAM :-)\n"
    "\n"
    "\t-K\t\tRun test mode\n"
    "\n"
    "\t-o ofile\tOutput formatted JSON to ofile (def: stdout, same as '-')\n"
    "\n"
    "\t-F fmt\t\tChange the JSON format style (def: use default)\n\n"
    "\t\t\ttty\t\tWhen output is to a TTY, use colour, otherwise use simple\n"
    "\t\t\tsimple\t\tEach line has one JSON level determined by '[]'s and '{}'s\n"
    "\t\t\tcolour\t\tSimple plus ANSI colour syntax highlighting\n"
    "\t\t\tcolor\t\tAlias for colour (colour excluding you :-) )\n"
    "\t\t\t1line\t\tOne line output\n"
    "\t\t\tnows\t\tOne line output, no extra whitespace\n"
    "\n"
    "\tfile.json\tJSON file to parse (- ==> read from stdin)\n"
    "\n"
    "Exit codes:\n"
    "    0\tall is OK and file is valid JSON\n"
    "    1\terror writing to ofile\n"
    "    2\t-h and help string printed or -V and version string printed\n"
    "    3\tinvalid command line, invalid option or option missing an argument\n"
    "    4\tfile does not exist, not a file, or unable to read the file\n"
    "    5\tfile contents is not valid JSON\n"
    "    6\ttest mode failed\n"
    " >=10\tinternal error\n"
    "\n"
    "JSON parser version: %s\n"
    "jfmt version: %s";

/*
 * functions
 */
static void usage(int exitcode, char const *prog, char const *str) __attribute__((noreturn));

/*
 * usage - print usage to stderr
 *
 * Example:
 *      usage(3, program, "wrong number of arguments");;
 *
 * given:
 *	exitcode        value to exit with
 *	str		top level usage message
 *	prog		our program name
 *
 * NOTE: We warn with extra newlines to help internal fault messages stand out.
 *       Normally one should NOT include newlines in warn messages.
 *
 * This function does not return.
 */
static void
usage(int exitcode, char const *prog, char const *str)
{
    /*
     * firewall
     */
    if (prog == NULL) {
	prog = "((NULL prog))";
	warn(__func__, "\nin usage(): prog was NULL, forcing it to be: %s\n", prog);
    }
    if (str == NULL) {
	str = "((NULL str))";
	warn(__func__, "\nin usage(): str was NULL, forcing it to be: %s\n", str);
    }

    /*
     * print the formatted usage stream
     */
    if (*str != '\0') {
	fprintf_usage(DO_NOT_EXIT, stderr, "%s\n", str);
    }
    fprintf_usage(DO_NOT_EXIT, stderr, usage_msg0, prog, DBG_DEFAULT, JSON_DBG_DEFAULT,
	    JSON_DEFAULT_MAX_DEPTH, JSON_PARSER_VERSION, JFMT_VERSION);
    exit(exitcode); /*ooo*/
    not_reached();
}


int
main(int argc, char **argv)
{
    char const *program = NULL;		/* our name */
    extern char *optarg;
    extern int optind;
    struct jfmt *jfmt = NULL;	/* struct of all our options and other things */
    size_t len = 0;			/* length of file contents */
    bool is_valid = false;		/* if file is valid json */
    int exit_code = 0;			/* for the end */
    int i;

    jfmt = alloc_jfmt();		/* allocate our struct jfmt * */
    /*
     * the alloc_jfmt() will never return a NULL pointer but check just in
     * case
     */
    if (jfmt == NULL) {
	err(23, "jfmt", "failed to allocate jfmt struct");
	not_reached();
    }

    /*
     * parse args
     */
    program = argv[0];
    while ((i = getopt(argc, argv, ":hVv:J:qL:I:l:m:Ko:F:")) != -1) {
	switch (i) {
	case 'h':		/* -h - print help to stderr and exit 0 */
	    free_jfmt(&jfmt);
	    usage(2, program, "");	/*ooo*/
	    not_reached();
	    break;
	case 'V':		/* -V - print version and exit */
	    free_jfmt(&jfmt);
	    print("%s\n", JFMT_VERSION);
	    exit(2);		/*ooo*/
	    not_reached();
	    break;
	case 'v':		/* -v verbosity */
	    /*
	     * parse verbosity
	     */
	    verbosity_level = parse_verbosity(optarg);
	    if (verbosity_level < 0) {
		usage(3, program, "invalid -v verbosity"); /*ooo*/
		not_reached();
	    }
	    break;
	case 'J':		/* -J json_verbosity */
	    /*
	     * parse JSON verbosity level
	     */
	    json_verbosity_level = parse_verbosity(optarg);
	    if (verbosity_level < 0) {
		usage(3, program, "invalid -J json_verbosity"); /*ooo*/
		not_reached();
	    }
	    break;
	case 'q':
	    quiet = true;
	    msg_warn_silent = true;
	    break;
	case 'L':
	    jfmt->common.print_json_levels = true; /* print JSON levels */
	    json_util_parse_st_level_option(optarg, &jfmt->common.num_level_spaces, &jfmt->common.print_level_tab);
	    break;
	case 'I':
	    jfmt->common.indent_levels = true;
	    json_util_parse_st_indent_option(optarg, &jfmt->common.indent_spaces, &jfmt->common.indent_tab);
	    break;
	case 'l':
	    jfmt->common.levels_constrained = true;
	    json_util_parse_number_range("-l", optarg, false, &jfmt->common.json_util_levels);
	    break;
	case 'm': /* set maximum depth to traverse json tree */
	    if (!string_to_uintmax(optarg, &jfmt->common.max_depth)) {
		free_jfmt(&jfmt);
		err(3, "jfmt", "couldn't parse -m depth"); /*ooo*/
		not_reached();
	    }
	    break;
	case 'K': /* run test code */
	    if (!jfmt_run_tests()) {
		free_jfmt(&jfmt);
		exit(6); /*ooo*/
	    }
	    else {
		free_jfmt(&jfmt);
		exit(0); /*ooo*/
	    }
	    break;
	case 'o': /* search with OR mode */
	    if (strcmp(optarg, "-")) { /* check if we will write to stdout */
		jfmt->common.outfile_not_stdout = true;
	    }
	    jfmt->common.outfile_path = optarg;
	    break;
	case 'F':
	    /*
	     * setting the common.format and common.format_output_changed is redundant
	     * but we do it anyway
	     */
	    jfmt->common.format = JSON_FMT_TTY; /* assume default */
	    jfmt->common.format = parse_json_util_format(&jfmt->common, "jfmt", optarg);
	    break;
	case ':':   /* option requires an argument */
	case '?':   /* illegal option */
	default:    /* anything else but should not actually happen */
	    check_invalid_option(program, i, optopt);
	    usage(3, program, ""); /*ooo*/
	    not_reached();
	    break;
	}
    }


    /*
     * check for conflicting options prior to changing argc and argv so that the
     * user will know to correct the options before being told that they have
     * the wrong number of arguments (if they do). Not everything can be checked
     * prior to doing this though.
     */

    /* run specific sanity checks on options etc. */
    jfmt->common.json_file = jfmt_sanity_chks(jfmt, program, &argc, &argv);


    /*
     * jfmt_sanity_chks() should never return a NULL FILE * but we check
     * anyway
     */
    if (jfmt->common.json_file == NULL) {
	/*
	 * NOTE: don't make this exit code 3 as it's an internal error if the
	 * jfmt_sanity_chks() returns a NULL pointer.
	 */
	err(24, "jfmt", "could not open regular readable file");
	not_reached();
    }

    /*
     * Read in contents of file.
     *
     * NOTE: why doesn't the jfmt_sanity_chks() function do this? Because this
     * is not so much about a sane environment as much as being unable to
     * continue after verify the command line is correct.
     */
    jfmt->common.file_contents = read_all(jfmt->common.json_file, &len);
    if (jfmt->common.file_contents == NULL) {
	err(4, "jfmt", "could not read in file: %s", argv[0]); /*ooo*/
	not_reached();
    }

    jfmt->common.json_tree = parse_json(jfmt->common.file_contents, strlen(jfmt->common.file_contents),
	    jfmt->common.json_file_path, &is_valid);
    if (!is_valid || jfmt->common.json_tree == NULL) {
	if (jfmt->common.json_file != stdin) {
	    fclose(jfmt->common.json_file);  /* close file prior to exiting */
	    jfmt->common.json_file = NULL;   /* set to NULL even though we're exiting as a safety precaution */
	}

	/* free our jfmt struct */
	free_jfmt(&jfmt);
	err(5, "jfmt", "%s invalid JSON", argv[0]); /*ooo*/
	not_reached();
    }

    dbg(DBG_MED, "valid JSON");

    /* only if we get here can we try and open the output file */
    if (jfmt->common.outfile_not_stdout && jfmt->common.outfile_path != NULL && *jfmt->common.outfile_path != '\0') {
	/*
	 * we know the output file doesn't exist but the real question is can we
	 * open it for writing to make it JSON ?
	 */
	jfmt->common.outfile = fopen(jfmt->common.outfile_path, "w");
	if (jfmt->common.outfile == NULL) {
	    err(1, __func__, "couldn't open output file: %s", jfmt->common.outfile_path);/*ooo*/
	    not_reached();
	}
    }

    /* XXX - change this to format the file - XXX */
    fpr(jfmt->common.outfile?jfmt->common.outfile:stdout, "jfmt", "%s", jfmt->common.file_contents);

    /* free tree */
    json_tree_free(jfmt->common.json_tree, jfmt->common.max_depth);

    /* All Done!!! -- Jessica Noll, Age 2 */
    if (jfmt != NULL) {
	free_jfmt(&jfmt);	/* free jfmt struct */
    }

    exit(exit_code); /*ooo*/
}

/* jfmt_sanity_chks	- sanity checks on jfmt tool options
 *
 * given:
 *
 *	jfmt	    - pointer to our jfmt struct
 *	program	    - program name
 *	argc	    - pointer to argc from main()
 *	argv	    - pointer to argv from main()
 *
 * This function runs any important checks on the jfmt internal state. It also
 * runs checks that a file arg is specified and that the right number of options
 * are specified done after options are parsed.
 *
 * If passed a NULL pointer or anything is not sane this function will not
 * return.
 *
 * This function returns a FILE *, the file to read the json from. It will not
 * return if this cannot be done (i.e. it will never return a NULL pointer
 * though main() still checks to be defensive).
 *
 * NOTE: this function does NOT check for valid JSON.
 *
 * NOTE: jfmt->check_tool_path and jfmt->check_tool_args can be NULL.
 *
 * NOTE: this function must be in jfmt.c, not jfmt_util.h, because it uses
 * the usage() function which needs to be in this file.
 */
FILE *
jfmt_sanity_chks(struct jfmt *jfmt, char const *program, int *argc, char ***argv)
{
    /* firewall */
    if (jfmt == NULL) {
	err(25, __func__, "NULL jfmt");
	not_reached();
    } else if (argc == NULL) {
	err(26, __func__, "NULL argc");
	not_reached();
    } else if (argv == NULL || *argv == NULL || **argv == NULL) {
	err(27, __func__, "NULL argv");
	not_reached();
    } else if (program == NULL) {
	err(28, __func__, "NULL program");
	not_reached();
    }

    /*
     * shift argc and argv for processing. They're a pointer to those in
     * main() so we have to dereference them here because main() also requires
     * that they are shifted.
     */
    (*argc) -= optind;
    (*argv) += optind;

    /* must have at least one arg */
    if ((*argv)[0] == NULL) {
	usage(3, program, "wrong number of arguments"); /*ooo*/
	not_reached();
    }
    /* check that file path is not an empty string */
    if (*(*argv)[0] == '\0') {
	usage(3, program, "empty file path");/*ooo*/
	not_reached();
    }

    /* if argv[0] != "-" we will attempt to open a regular readable file */
    if (strcmp((*argv)[0], "-") != 0) {
	/*
	 * first check that if -o was used that the out file is not the same as
	 * the file.json file! Do this in a case-insensitive way for file
	 * systems that don't distinguish file.json from FILE.JSON.
	 *
	 * NOTE: do NOT open this file: we won't do that until we know that the
	 * input JSON is valid.
	 */
	if (jfmt->common.outfile_not_stdout && jfmt->common.outfile_path != NULL && *jfmt->common.outfile_path != '\0') {
	    if (!strcasecmp((*argv)[0], jfmt->common.outfile_path)) {
		free_jfmt(&jfmt);
		err(3, __func__, "-o ofile is the same as JSON file"); /*ooo*/
		not_reached();
	    }
	}
        /* check that the path exists and is a regular readable file */
	if (!exists((*argv)[0])) {
	    free_jfmt(&jfmt);
	    err(4, __func__, "%s: file does not exist", (*argv)[0]); /*ooo*/
	    not_reached();
	} else if (!is_file((*argv)[0])) {
	    free_jfmt(&jfmt);
	    err(4, __func__, "%s: not a regular file", (*argv)[0]); /*ooo*/
	    not_reached();
	} else if (!is_read((*argv)[0])) {
	    free_jfmt(&jfmt);
	    err(4, __func__, "%s: unreadable file", (*argv[0])); /*ooo*/
	    not_reached();
	}

	errno = 0; /* pre-clear errno for errp() */
	jfmt->common.json_file_path = (*argv)[0];
	jfmt->common.json_file = fopen(jfmt->common.json_file_path, "r");
	if (jfmt->common.json_file == NULL) {
	    free_jfmt(&jfmt);
	    errp(4, __func__, "%s: could not open for reading", (*argv)[0]); /*ooo*/
	    not_reached();
	}
    } else { /* argv[0] is "-": will read from stdin */
	jfmt->common.json_file_path = "-";
	jfmt->common.is_stdin = true;
	jfmt->common.json_file = stdin;
    }

    /*
     * if -o file specified and not stdout (not '-') then we have to verify that
     * it does not already exist and that we can open it for writing
     *
     * NOTE: there is a slight risk that between the time we have checked if the
     * file exists and when we want to open it to write it later that it exists
     */
    if (jfmt->common.outfile_not_stdout && jfmt->common.outfile_path != NULL && *jfmt->common.outfile_path != '\0') {
	if (exists(jfmt->common.outfile_path)) {
	    err(3, __func__, "-o ofile already exists: will not overwrite"); /*ooo*/
	    not_reached();
	}
    }

    dbg(DBG_LOW, "maximum depth to traverse: %ju%s", jfmt->common.max_depth, (jfmt->common.max_depth == 0?" (no limit)":
		jfmt->common.max_depth==JSON_DEFAULT_MAX_DEPTH?" (default)":""));


    /* XXX - should it be an error if any more args are specified ? - XXX */

    /* all good: return the (presumably) json FILE * */
    return jfmt->common.json_file;
}
